File: C++/edge.cpp
#include "edge.h"
#include <cmath>
#include <vector>
#include <stdexcept>


template <typename T>
double mean(const std::vector<T> &x){
  unsigned int n = x.size(); double sum = 0.0;
  for(unsigned int i=0; i<n; i++) sum += x[i];
  return sum / n;
}


double edge(
    const std::vector<double> &open,
    const std::vector<double> &high,
    const std::vector<double> &low,
    const std::vector<double> &close,
    const bool sign){

  unsigned int n = open.size();
  if(high.size() != n or low.size() != n or close.size() != n){
    throw std::invalid_argument(
        "open, high, low, close must have the same length"
    );
  }
  
  std::vector<double> o(n), h(n), l(n), c(n), m(n);
  for(unsigned int i=0; i<n; i++){
    o[i] = std::log(open[i]);
    h[i] = std::log(high[i]);
    l[i] = std::log(low[i]);
    c[i] = std::log(close[i]);
    m[i] = (h[i] + l[i]) / 2.;
  }
  
  std::vector<unsigned int> tau(n-1), phi1(n-1), phi2(n-1), phi3(n-1), phi4(n-1);
  std::vector<double> r1(n-1), r2(n-1), r3(n-1), r4(n-1), r5(n-1);
  for(unsigned int i=0; i<n-1; i++){
    tau[i] = (h[i+1] != l[i+1]) | (l[i+1] != c[i]);
    phi1[i] = (o[i+1] != h[i+1]) & tau[i];
    phi2[i] = (o[i+1] != l[i+1]) & tau[i];
    phi3[i] = (c[i] != h[i]) & tau[i];
    phi4[i] = (c[i] != l[i]) & tau[i];
    r1[i] = m[i+1] - o[i+1];
    r2[i] = o[i+1] - m[i];
    r3[i] = m[i+1] - c[i];
    r4[i] = c[i] - m[i];
    r5[i] = o[i+1] - c[i];
  }
  
  double pt = mean(tau);
  double m1 = mean(r1), m3 = mean(r3), m5 = mean(r5); 
  std::vector<double> d1(n-1), d3(n-1), d5(n-1);
  for(unsigned int i=0; i<n-1; i++){
    d1[i] = r1[i] - tau[i] * m1 / pt;
    d3[i] = r3[i] - tau[i] * m3 / pt;
    d5[i] = r5[i] - tau[i] * m5 / pt;
  }
  
  double po = mean(phi1) + mean(phi2), pc = mean(phi3) + mean(phi4);
  std::vector<double> x1(n-1), x2(n-1), x11(n-1), x22(n-1);
  for(unsigned int i=0; i<n-1; i++){
    x1[i] = -4./po*d1[i]*r2[i] -4./pc*d3[i]*r4[i];
    x2[i] = -4./po*d1[i]*r5[i] -4./pc*d5[i]*r4[i];
    x11[i] = x1[i] * x1[i];
    x22[i] = x2[i] * x2[i];
  }
  
  double e1 = mean(x1), e2 = mean(x2);
  double v1 = mean(x11) - e1*e1, v2 = mean(x22) - e2*e2;
  double s2 = (v2*e1 + v1*e2) / (v1 + v2);
  double s = std::sqrt(std::abs(s2));
  
  if(sign & (s2 < 0))
    s = -s;
  
  return s;
  
}

File: python/bidask/edge.py
import warnings
import numpy as np


def edge(open: np.array, high: np.array, low: np.array, close: np.array, sign: bool = False) -> float:
    """
    Efficient Estimation of Bid-Ask Spreads from Open, High, Low, and Close Prices

    Implements the efficient estimator of bid-ask spreads from open, high, low, 
    and close prices described in Ardia, Guidotti, & Kroencke (JFE, 2024):
    https://doi.org/10.1016/j.jfineco.2024.103916

    Parameters
    ----------
    - `open`: array-like 
        Vector of open prices sorted in ascending order of the timestamp.
    - `high`: array-like 
        Vector of high prices sorted in ascending order of the timestamp.
    - `low`: array-like 
        Vector of low prices sorted in ascending order of the timestamp.
    - `close`: array-like 
        Vector of close prices sorted in ascending order of the timestamp.
    - `sign`: 
        Whether to return signed estimates.

    Returns
    -------
    float
        The spread estimate. A value of 0.01 corresponds to a spread of 1%.
    """
    # check that the open, high, low, and close prices have the same length
    nobs = len(open)
    if len(high) != nobs or len(low) != nobs or len(close) != nobs:
        raise ValueError("Open, high, low, and close prices must have the same length")

    # return missing if there are less than 3 observations
    if nobs < 3:
        return np.nan

    # compute log-prices
    o = np.log(np.asarray(open))
    h = np.log(np.asarray(high))
    l = np.log(np.asarray(low))
    c = np.log(np.asarray(close))
    m = (h + l) / 2.

    # shift log-prices by one period
    h1, l1, c1, m1 = h[:-1], l[:-1], c[:-1], m[:-1]
    o, h, l, c, m = o[1:], h[1:], l[1:], c[1:], m[1:]

    # compute log-returns
    r1 = m - o
    r2 = o - m1
    r3 = m - c1
    r4 = c1 - m1
    r5 = o - c1

    # compute indicator variables
    tau = np.where(np.isnan(h) | np.isnan(l) | np.isnan(c1), np.nan, (h != l) | (l != c1))
    po1 = tau * np.where(np.isnan(o) | np.isnan(h), np.nan, o != h)
    po2 = tau * np.where(np.isnan(o) | np.isnan(l), np.nan, o != l)
    pc1 = tau * np.where(np.isnan(c1) | np.isnan(h1), np.nan, c1 != h1)
    pc2 = tau * np.where(np.isnan(c1) | np.isnan(l1), np.nan, c1 != l1)
    
    # ignore warnings raised by nanmean for all-NaN slices
    with warnings.catch_warnings():
        warnings.simplefilter('ignore', RuntimeWarning)

        # compute probabilities
        pt = np.nanmean(tau)
        po = np.nanmean(po1) + np.nanmean(po2)
        pc = np.nanmean(pc1) + np.nanmean(pc2)

        # return missing if there are less than two periods with tau=1 
        # or po or pc is zero
        if np.nansum(tau) < 2 or po == 0 or pc == 0:
            return np.nan
    
        # compute de-meaned log-returns
        d1 = r1 - np.nanmean(r1)/pt*tau
        d3 = r3 - np.nanmean(r3)/pt*tau
        d5 = r5 - np.nanmean(r5)/pt*tau
    
        # compute input vectors
        x1 = -4./po*d1*r2 + -4./pc*d3*r4 
        x2 = -4./po*d1*r5 + -4./pc*d5*r4 
    
        # compute expectations
        e1 = np.nanmean(x1)
        e2 = np.nanmean(x2)
    
        # compute variances
        v1 = np.nanmean(x1**2) - e1**2
        v2 = np.nanmean(x2**2) - e2**2
    
    # compute square spread by using a (equally) weighted 
    # average if the total variance is (not) positive
    vt = v1 + v2
    s2 = (v2*e1 + v1*e2) / vt if vt > 0 else (e1 + e2) / 2.
    
    # compute signed root
    s = np.sqrt(np.abs(s2))
    if sign: 
        s *= np.sign(s2)
  
    # return the spread
    return float(s)
